### DISCUSSION TITLE
Idiomatic way to handle `Pytree` flatten/unflatten at `jax.Primitive` boundaries

### DISCUSSION THREAD

**Discussion by femtomc** [jax-ml/jax#13574] [Q&A] [answered]
I've defined new primitives to use - but I've run into a recurring issue, which I've solved using something which I think may be hacky.

The issue is that I have primitives that I call `bind` with, and sometimes these primitives are given `Pytree` instances - not valid JAX types.

Normally, I'll just flatten the trees as the `bind` boundary -- and then give the primitive a `PyTreeDef` as a static value so that it can unflatten the arguments on the other side of the boundary when I give it semantics (by dispatching with it, using another mechanism).

Is this generally the right thing to do here? Am I paying some repeated cost during tracing, which I might avoid?

Thanks for any help.

**Comment by jakevdp** [comment, marked as answer]
I'd suggest defining your primitive such that they only ever take arrays as arguments. Then you can create a wrapper function around those primitives that can flatten those inputs when necessary. Here's an example of what I have in mind: in `jax.experimental.sparse`, we define the `bcoo_todense` primitive that operates on arrays. Its high-level wrapper is `bcoo_todense()` which accepts a `BCOO` object (a pytree) and extracts the underlying arrays to pass to the primitive: https://github.com/google/jax/blob/2c92037150f0eaa767ab1f663296ee1c7338a48e/jax/experimental/sparse/bcoo.py#L172-L196

It doesn't directly call pytree.flatten, but essentially pulls out the same information, and the key is that the `bind()` call will never see anything but arrays as inputs. Does that match the kind of situation you have in mind?

**Reply by femtomc** [reply]
Thanks Jake -- this is great.
